### PR TITLE
CPU reductions - mean and standard deviation, with tree reduction.

### PR DESCRIPTION
**Pull request by mzient**
Signed-off-by: Michal Zientkiewicz <michalz@nvidia.com>

#### Why we need this PR?
*Pick one, remove the rest*
- It adds new feature needed for Normalize operator

#### What happened in this PR?
*Fill relevant points, put NA otherwise. Replace anything inside []*
 - What solution was applied:
     * Implemented reductions (on CPU) using pairwise reduction model to improve numerical stability
     * minor improvements in common parts of the library
 - Affected modules and functionalities:
     * kernels library (added the reductions)
     * core library (tensor manipulation, SmallVector warning)
 - Key points relevant for the review:
     * reduce.h and tests
 - Validation and testing:
     * Unit tests (GTEST)
 - Documentation (including examples):
     * Doxygen

**JIRA TASK**: DALI-1242

### CODE REVIEW

**Comment by mzient** [comment]
!build

File: include/dali/core/small_vector.h [hunk 1]
```diff
@@ -180,7 +180,7 @@ class SmallVector : SmallVectorAlloc<T, allocator>, SmallVectorBase<T> {
     clear();
     reserve(n);
     T *ptr = data();
-    for (size_t i = 0; i < n; i++) {
+    for (decltype(n) i = 0; i < n; i++) {
```
Comment (mzient) [inline include/dali/core/small_vector.h]: silence signed/unsigned mismatch warning when converting from a `span`, whose size() returns ptrdiff_t.

**Comment by mzient** [comment]
!build

**Comment by mzient** [comment]
!build

File: include/dali/core/tensor_view.h [hunk 2]
```diff
@@ -671,6 +671,24 @@ subtensor(TensorView<StorageBackend, DataType, DynamicDimensions> source, int64_
  * @}
  */
 
+/**
+ * @brief Merges a dimension with the next one
+ *
+ * @details The output tensor represents the same data, but the shape has two dimensions
+ * collapsed into one, e.g. a tensor with shape
+ * [2, 3, 4, 5]
+ * after a call to flatten(tensor, 1) would have a shape:
```
Comment (JanuszL) [inline include/dali/core/tensor_view.h]:
```suggestion
 * after a call to collapse_dim(tensor, 1) would have a shape:
```

File: include/dali/core/tensor_shape.h [hunk 3]
```diff
@@ -1049,6 +1050,32 @@ TensorListShape<ndim> uniform_list_shape(int num_samples, std::initializer_list<
   return TensorListShape<ndim>::make_uniform(num_samples, sample_shape);
 }
 
+/**
+ * @brief Merges a dimension with the next one
+ *
+ * @details The output shape has two specified dimension collapsed with the next one
+ * [2, 3, 4, 5]
+ * after a call to flatten(tensor, 1) would have a shape:
```
Comment (JanuszL) [inline include/dali/core/tensor_shape.h]:
```suggestion
 * after a call to collapse_dim(tensor, 1) would have a shape:
```

File: include/dali/core/tensor_shape.h [hunk 4]
```diff
@@ -1049,6 +1050,32 @@ TensorListShape<ndim> uniform_list_shape(int num_samples, std::initializer_list<
...
+ *
+ * @param dim_idx - the dimension to drop; must be >= 0 and < shape.size() - 1
+ * @param rv - input TensorView
+ * @remarks The `dim_idx` must not be the innermost dimension or the result is undefined.
+ */
+DALI_NO_EXEC_CHECK
+template <int ndim>
+DALI_HOST_DEV
+auto collapse_dim(const TensorShape<ndim> &shape, int dim_idx) {
+  TensorShape<(ndim > 0 ? ndim-1 : -1)> ret;
```
Comment (JanuszL) [inline include/dali/core/tensor_shape.h]: Why can't we just assert:
```
assert(dim_idx >= 0 && dim_idx < shape.size() - 1)
```

File: dali/kernels/reduce/reduce.h [hunk 5]
```diff
@@ -0,0 +1,417 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
```
Comment (JanuszL) [inline dali/kernels/reduce/reduce.h]:
```suggestion
// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
```

File: dali/kernels/reduce/reduce_test.cc [hunk 6]
```diff
@@ -0,0 +1,246 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
```
Comment (JanuszL) [inline dali/kernels/reduce/reduce_test.cc]:
```suggestion
// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
```

File: dali/kernels/reduce/reduce_test.cc [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+
+#include <gtest/gtest.h>
+#include <random>
+#include <chrono>
+#include "dali/kernels/reduce/reduce.h"
+
+namespace dali {
+namespace kernels {
+
+using perfclock = std::chrono::high_resolution_clock;
```
Comment (JanuszL) [inline dali/kernels/reduce/reduce_test.cc]: Is it only for debugging? Maybe perf test should be moved to the benchmark?

File: dali/kernels/reduce/reduce.h [hunk 8]
```diff
@@ -0,0 +1,417 @@
...
+  int64_t stride = in.stride[axis];
+  if (axis == in.dim() - 1) {
+    Dst tmp = 0;
+    reduce1D(tmp, in.data + offset, stride, in.size[axis], P, R);
+    R(reduced, tmp);
+  } else {
+    int64_t sub_v = volume(in.size.begin() + axis + 1, in.size.end());
+    if (!sub_v)
+      sub_v = 1;
+    if (extent >= 4 && extent * sub_v > kTreeReduceThreshold) {
```
[the remark this answers is not in the document]
Comment (mzient) [reply dali/kernels/reduce/reduce.h]: It was an optimization, but maybe it's not worth it... In that case, >= 2. I'll check the perf.

File: dali/kernels/reduce/reduce.h [hunk 9]
```diff
@@ -0,0 +1,417 @@
...
+             const InTensorCPU<Src, -1> &in,
+             span<const int> axes) {
+    if (in.shape.size() > 64)
+      throw std::range_error("Reduce supports up to 64 dimensions");
+    input = in;
+    output = out;
+    setup_axes(axes);
+    setup_output();
+    setup_input();
+    assert(output.num_elements() == out.num_elements());
```
[the remark this answers is not in the document]
Comment (mzient) [reply dali/kernels/reduce/reduce.h]: No, output is the product of our setup.
What I should enforce - even before we setup anything - is that out has a proper reduced shape (with reduced dims dropped or forced to 1) - now I'm just checking that the number of elements matches, but that may also mean permuted dims or sth.

File: dali/kernels/reduce/reduce.h [hunk 10]
```diff
@@ -0,0 +1,417 @@
...
+  uint64_t operator()(uint32_t x) const noexcept {
+    return static_cast<uint64_t>(x) * x;
+  }
+};
+
+template <typename Mean>
+struct variance {
+  Mean mean = 0;
+  template <typename T>
+  auto operator()(const T &x) const noexcept {
```
[the remark this answers is not in the document]
Comment (mzient) [reply dali/kernels/reduce/reduce.h]: It's already promoted to the result of mean subtraction. Promotion would only make sense if mean is an integer.

**Comment by mzient** [comment]
!build

File: dali/kernels/reduce/reduce_test.cc (hunk 7 above)
Comment (mzient) [reply dali/kernels/reduce/reduce_test.cc]: Sort of. It doesn't spam the output too much and when we have time (I wonder when, hehe) to optimize, then it's already there...

**Comment by mzient** [comment]
!build

File: dali/kernels/reduce/reduce.h [hunk 11]
```diff
@@ -0,0 +1,417 @@
...
+    auto d = x - mean;
+    return d * d;
+  }
+};
+
+struct sum {
+  template <typename Acc, typename Addend>
+  DALI_HOST_DEV DALI_FORCEINLINE
```
[the remark this answers is not in the document]
Comment (mzient) [reply dali/kernels/reduce/reduce.h]: Done

File: include/dali/core/tensor_view.h [hunk 12]
```diff
@@ -671,6 +671,24 @@ subtensor(TensorView<StorageBackend, DataType, DynamicDimensions> source, int64_
...
+ * [2, 12, 5]
+ *
+ * @param dim_idx - the dimension to drop; must be >= 0 and < tv.dim() - 1
+ * @param rv - input TensorView
+ * @remarks The `dim_idx` must not be the innermost dimension or the result is undefined.
```
[the remark this answers is not in the document]
Comment (mzient) [reply include/dali/core/tensor_view.h]: That would only be possible for dynamically sized TensorShape - for static one the return type demands that I drop one dimension - and I don't want this behavior to depend on whether the shape is dynamic or not.

File: include/dali/core/tensor_shape.h (hunk 4 above)
Comment (mzient) [reply include/dali/core/tensor_shape.h]: Done.

[the remark this answers is not in the document]
Comment (mzient) [reply include/dali/core/tensor_shape.h]: Done.

**Comment by mzient** [comment]
!build

File: dali/kernels/reduce/reduce.h [hunk 13]
```diff
@@ -0,0 +1,453 @@
...
+#include "dali/core/static_switch.h"
+#include "dali/core/tensor_shape_print.h"
+#include "dali/core/tensor_view.h"
+#include "dali/core/util.h"
+
+namespace dali {
+namespace kernels {
+namespace reductions {
+
+struct identity {
```
Comment (szalpal) [inline dali/kernels/reduce/reduce.h]: These shall be upper-case https://google.github.io/styleguide/cppguide.html#Type_Names

File: dali/kernels/reduce/reduce.h [hunk 14]
```diff
@@ -0,0 +1,453 @@
...
+  }
+};
+
+}  // namespace reductions
+
+
+constexpr int kTreeReduceThreshold = 32;
+
+template <int static_stride, typename Dst, typename Src, typename Preprocessor, typename Reduction>
+void reduce1D_stride(Dst &reduced, const Src *data, int64_t dynamic_stride, int64_t n,
```
Comment (szalpal) [inline dali/kernels/reduce/reduce.h]: These shall be upper-case https://google.github.io/styleguide/cppguide.html#Function_Names

File: dali/kernels/reduce/reduce.h [hunk 15]
```diff
@@ -0,0 +1,453 @@
...
+      for (int64_t i = 0; i < extent; i++) {
+        Dst tmp = 0;
+        reduce(tmp, in, P, R, axis + 1, in.size[axis + 1], offset + i * stride);
+        R(reduced, tmp);
+      }
+    }
+  }
+}
+
+/// @brief Reduces a strided tensor to a scalar
```
Comment (szalpal) [inline dali/kernels/reduce/reduce.h]: I thought we agreed on a documentation style... https://github.com/NVIDIA/DALI/blob/master/STYLE_GUIDE.md#documentation

Comment (mzient) [reply dali/kernels/reduce/reduce.h]: I haven't. I was on vacation when that happened.

**Comment by mzient** [comment]
!build

File: include/dali/core/tensor_shape.h [hunk 16]
```diff
@@ -1049,6 +1050,34 @@ TensorListShape<ndim> uniform_list_shape(int num_samples, std::initializer_list<
...
+ * @brief Merges a dimension with the next one
+ *
+ * @details The output shape has two specified dimension collapsed with the next one
+ * [2, 3, 4, 5]
+ * after a call to collapse_dim(tensor, 1) would have a shape:
+ * [2, 12, 5]
+ *
+ * @param dim_idx - the dimension to drop; must be >= 0 and < shape.size() - 1
+ * @param rv - input TensorView
+ * @remarks The `dim_idx` must not be the innermost dimension or the result is undefined.
```
Comment (JanuszL) [inline include/dali/core/tensor_shape.h]: This remark should say that it would just assert.

File: dali/kernels/reduce/reduce.h (hunk 15 above)
Comment (JanuszL) [reply dali/kernels/reduce/reduce.h]: Still, majority has voted and you should submit.

File: include/dali/core/tensor_shape.h (hunk 16 above)
Comment (mzient) [reply include/dali/core/tensor_shape.h]: It's only true in debug build.

File: dali/kernels/reduce/reduce.h (hunk 15 above)
Comment (mzient) [reply dali/kernels/reduce/reduce.h]: I'm more than willing to submit - a change to that guide.

**Pull request merged**